### PR TITLE
forceResize parameter for thumbnail processors

### PR DESCRIPTION
**Pull request by Yivan**
Pull request for issue #786

It is a corrected version of PR https://github.com/pimcore/pimcore/pull/835

It includes:
- for cover processor "doNotScaleUp" has been changed to "forceResize" for consistency accross the code.
- Implemented for all resizing methods wich could need this: scaleByWidth, scaleByHeight, cover, contain and frame
- Could be used as inline thumbnail generation: `$assetImage->getThumbnail(["frame" => true, "width" => 500, "height" => 300, "forceResize" => true]);`
- GUI Thumbnail on Pimcore backend updated for extjs4 and extjs6 so it can be configured and used from here
- Bug corrected for cover: in GUI by default "Positioning" is empty, so in the method "cover" of Adapter.php, "center" is set as default if empty (as it is done actually in getByLegacyConfig method of Config object)

I think you need to create maybe a post-update script to change thumbnail preset using cover (because it was doNotScaleUp before this PR). Maybe in the upgrade note, it should be said that if doNotScaleUp was used for inline thumbnail generation code need to be adapted. Ex:
`$assetImage->getThumbnail(["cover" => true, "width" => 500, "height" => 300, "doNotScaleUp" => false]);`
Should be converted to:
`$assetImage->getThumbnail(["cover" => true, "width" => 500, "height" => 300, "forceResize" => true]);`

### CODE REVIEW

**Comment by Yivan** [comment]
A translation need to be added: t("force_resize")

**Comment by brusch** [comment]
Translations are added in `pimcore/config/texts/en.json`

**Comment by Yivan** [comment]
I am really sorry to come back on this PR, and i don't like to do this, so i hope not bothering you with this.

But actually on many project we need to "patch" core pimcore files to be able to resize asset more than their original size. Without this it is impossible to use Pimcore asset system to upsizing.

If our PR has some problem feel free to ask modification, i will update it with pleasure.

If this is internal decision waiting, maybe you could at least for the moment just modify two lines in the core by replacing here https://github.com/pimcore/pimcore/blob/master/pimcore/models/Asset/Image/Thumbnail/Processor.php#L29:
```
"scaleByWidth" => ["width"],
"scaleByHeight" => ["height"],
```
by:
```
"scaleByWidth" => ["width", "forceResize"],
"scaleByHeight" => ["height", "forceResize"],
```

This will prevent forceResize to be filtered and unusable with the asset API (see filter here: https://github.com/pimcore/pimcore/blob/master/pimcore/models/Asset/Image/Thumbnail/Processor.php#L283).
By changing those 2 lines, we can now extend image getThumbnail and use "items" to construct a config for upsizing like:
```
$item = [ "method" => "scaleByWidth",
    "arguments" => [
        "width" => $config["width"],
        "forceResize" => true
    ]
];
...
$asset->getThumbnail(["items" => [$item]]);
```

I think it won't change any backward compatibility or introduce any problem. And it will be possible to upsizing asset for people needing it.

Thank you.

**Comment by Yivan** [comment]
I come back to this pending pull request.
Could I make any change to my PR so it can be accepted ? Do i need to update the code ? I will be happy to make any change as needed.
We really need to be able to generate thumbnail at the size wanted (even if upscaling), not being limited to the original resolution as maximum thumbnail resolution (for bitmap).

If not, are there any way to do it please ? (using thumbnail API or DI don't seem to work as explained in this issue).

Thanks for your feedback.

**Comment by elburro1887** [comment]
Please integrate this PR. I am also very much in need of this feature.

File: pimcore/lib/Pimcore/Image/Adapter.php [hunk 1]
```diff
@@ -185,18 +188,18 @@ public function contain($width, $height)
...
      * @param string $orientation
+     * @param  bool $forceResize
      * @return self
      */
-    public function cover($width, $height, $orientation = "center", $doNotScaleUp = true)
+    public function cover($width, $height, $orientation = "center", $forceResize = false)
     {
-        $scaleUp = $doNotScaleUp ? false : true;
-
+        if (empty($orientation)) $orientation = "center"; // if not set (from GUI for instance) - default value in getByLegacyConfig method of Config object too
```
Comment (maff) [inline pimcore/lib/Pimcore/Image/Adapter.php]: Could you please change this line to include braces (PSR-1/2)?

File: pimcore/lib/Pimcore/Image/Adapter/Imagick.php [hunk 2]
```diff
@@ -436,13 +436,14 @@ public function crop($x, $y, $width, $height)
     /**
      * @param $width
      * @param $height
+     * @param  bool $forceResize
      * @return $this
      */
-    public function frame($width, $height)
+    public function frame($width, $height, $forceResize)
```
Comment (maff) [inline pimcore/lib/Pimcore/Image/Adapter/Imagick.php]: Please add a default value for the forceResize param (BC)

**Comment by Yivan** [comment]
Thanks maff for your review.

I made all the changes requested, and add changes to ImageMagick adapter too. Postscript has been added and tested too. I hope all is ok.

The pull request is complete, so you can merge it.

Thanks a lot.

**Comment by Yivan** [comment]
(I was thinking when i readed Brush message a few weeks ago that he added the english translation, but reading it again today he was asking me add them and say to me how to do it, sorry my bad english. So I have added added force_resize translation as indicated).

**Comment by maff** [comment]
Looks good, thanks!

**Comment by Yivan** [comment]
@maff Thanks a lot, it was a so much needed one!

**Pull request merged**